### PR TITLE
[MXNET-780] Fix exception handling bug

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##
Exceptions thrown in OMP Parallel Blocks in ImageDetRecordIOParser and ImageRecordIOParser are not saved and transported to the main thread. This leads to a call to std::terminate. Fixing this by saving the exception and rethrowing from the main thread.

Depends on https://github.com/dmlc/dmlc-core/pull/445

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Fix Exception Handling Bug in ImageDetRecordIOParser
- [X] Fix Exception Handling Bug in ImageRecordIOParser

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/io/iter_image_det_recordio.cc [hunk 1]
```diff
@@ -237,53 +237,53 @@ class ImageDetRecordIOParser {
...
   mshadow::TensorContainer<cpu, 3> img_;
+  /*! \brief OMPException obj to store and rethrow exceptions from omp blocks*/
+  dmlc::OMPException omp_exc_;
 };
 
-template<typename DType>
+template <typename DType>
 inline void ImageDetRecordIOParser<DType>::Init(
-    const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    const std::vector<std::pair<std::string, std::string>> &kwargs) {
```
Comment (zhreshold) [inline src/io/iter_image_det_recordio.cc]: The space is left deliberately for compatibility with some compilers, don't remove it.

**Comment by zhreshold** [comment]
Nice fix to address the error handling problem. Related #12280

BTW, I think ultimately we have to catch all exceptions (MXNetError or DMLC error) in python so interactive session will no longer crash by segmentation fault or something.

**Comment by anirudh2290** [comment]
@zhreshold interactive session wont crash after this change but will throw an MXNetError.

**Comment by zhreshold** [comment]
@anirudh2290 I mean, we need to catch all exceptions. For example:

```python3
import mxnet as mx
a = mx.nd.zeros((1,2))
mx.nd.smooth_l1(a)
```
crash with 
```bash
libc++abi.dylib: terminating with uncaught exception of type std::invalid_argument: stod: no conversion
Abort trap: 6
```

**Comment by stu1130** [comment]
@anirudh2290 Could you please address the feedback. Thanks

File: src/io/iter_image_det_recordio.cc (hunk 1 above)
Comment (anirudh2290) [reply src/io/iter_image_det_recordio.cc]: I have added it back. Which compiler is it ? Maybe it should be added to the CI.

**Comment by vandanavk** [comment]
@anirudh2290 Please have a look at the CI build failure. Is this ready for review again once the CI is fixed?

**Comment by zhreshold** [comment]
LGTM

**Pull request merged**